src/v/archival/ntp_archiver_service.cc

```diff
@@ -1383,20 +1384,44 @@ ntp_archiver::schedule_uploads(std::vector<upload_context> loop_contexts) {
             _probe->upload_lag(ctx.last_offset - ctx.start_offset);
         }
 
-        while (uploads_remaining > 0 && may_begin_uploads()) {
-            auto should_stop = co_await ctx.schedule_single_upload(*this);
-            if (should_stop == ss::stop_iteration::yes) {
-                break;
-            }
+        std::exception_ptr ep;
+        try {
+            while (uploads_remaining > 0 && may_begin_uploads()) {
+                auto should_stop = co_await ctx.schedule_single_upload(*this);
+                if (should_stop == ss::stop_iteration::yes) {
+                    break;
+                }
 
-            // Decrement remaining upload count if the last call actually
-            // scheduled an upload.
-            if (!ctx.uploads.empty()) {
-                const auto& last_scheduled = ctx.uploads.back();
-                if (last_scheduled.result.has_value()) {
-                    uploads_remaining -= 1;
+                // Decrement remaining upload count if the last call actually
+                // scheduled an upload.
+                if (!ctx.uploads.empty()) {
+                    const auto& last_scheduled = ctx.uploads.back();
+                    if (last_scheduled.result.has_value()) {
+                        uploads_remaining -= 1;
+                    }
+                }
+            }
+        } catch (...) {
+            ep = std::current_exception();
+        }
+        if (ep) {
+            _rtclog.warn("Failed to schedule upload: {}", ep);
+            std::vector<ss::future<>> pending_uploads;
+            pending_uploads.reserve(ctx.uploads.size());
+            for (auto& upload : ctx.uploads) {
+                if (upload.result.has_value()) {
+                    pending_uploads.emplace_back(
+                      std::move(upload.result.value()).discard_result());
+                }
+            }
+            auto futs = co_await ss::when_all(
+              pending_uploads.begin(), pending_uploads.end());
+            for (auto& f : futs) {
+                if (f.failed()) {
+                    _rtclog.warn("Upload failed: {}", f.get_exception());
                 }
             }
+            std::rethrow_exception(ep);
```

Comment:
before we throw and return from this coroutine, do we also need to deal with the futures in `ctx.upload` for all the `ctx` in the input `std::vector<upload_context> loop_contexts` -- unless i'm missing something it looks like this new code handles the ctx for the current loop iteration?

Reply:
Oy vey! Great catch, we also need to mess with the ones moved out at the end of each loop iteration.

I'm really wanting to do a larger change here to fix the flow here - it's very complex and error prone. We don't have to check the rest of the loop_contexts in practice because they only get called in one place as parameters. Really this method should just take an "upload params" sort of struct and then kick off as many as it can and stop. All uploads should be aggregated into a single output vector so there isn't this moving of data out from the context into this local variable...

Reply:
Okay I simplified this a bunch IMO - PTAL